pkg/model/multiversion/manager.go

```diff
@@ -0,0 +1,197 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package multiversion
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+
+	"gopkg.in/src-d/go-git.v4"
+
+	ackgenconfig "github.com/aws-controllers-k8s/code-generator/pkg/generate/config"
+	ackmodel "github.com/aws-controllers-k8s/code-generator/pkg/model"
+	"github.com/aws-controllers-k8s/code-generator/pkg/util"
+)
+
+var (
+	ErrAPIVersionNotFound   = errors.New("api version not found")
+	ErrAPIVersionRemoved    = errors.New("api version removed")
+	ErrAPIVersionDeprecated = errors.New("api version deprecated")
+)
+
+// APIVersionManager is a API versions manager. It contains the mapping
+// of each non-deprecated version with their correspending ackmodel.Model
+// and APIInfos.
+type APIVersionManager struct {
+	gitRepo *git.Repository
+
+	hubVersion    string
+	spokeVersions []string
+
+	deprecatedVersions []string
+	removedVersions    []string
+
+	apiInfos map[string]APIInfo
+	models   map[string]*ackmodel.Model
+}
+
+// NewAPIVersionManager initialises and returns a new APIVersionManager.
+func NewAPIVersionManager(
+	sdkCacheDir string,
+	serviceAlias string,
+	hubVersion string,
+	apisInfo map[string]APIInfo,
+	defaultConfig ackgenconfig.Config,
+) (*APIVersionManager, error) {
+	if len(apisInfo) == 0 {
+		return nil, fmt.Errorf("empty apisInfo")
+	}
+
+	spokeVersions := make([]string, 0, len(apisInfo)-1)
+	gitRepo, err := util.LoadRepository(sdkCacheDir)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read sdk git repository: %v", err)
+	}
+
+	SDKAPIHelper := ackmodel.NewSDKHelper(sdkCacheDir)
+
+	// create model for each non-deprecated api version
+	models := make(map[string]*ackmodel.Model, len(apisInfo))
+	for apiVersion, apiInfo := range apisInfo {
+		// TODO(a-hilaly) filter deprecated and removed api versions
```

Comment:
I think we will continue to support all released versions of the API? Why would we ever deprecate an API version? It costs us nothing to generate the webhooks.

Reply:
We should continue to generate webhooks for `Deprecated` versions but not for the removed ones.
I think that at one point we should start deprecating versions and eventually remove them. I imagine that if one day we release a `v1` api version - all alpha/beta version should be at least "Deprecated" if not removed

Reply:
According to https://kubernetes.io/docs/reference/using-api/deprecation-policy/ we should not have more than one "Available" version at onces - and at most two "Deprecated" versions